oauth2l-go/src/oauth2client/oauth2client.go

```diff
@@ -0,0 +1,313 @@
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read http body for verification request.")
+	}
+	var token *Token
+	if err := json.Unmarshal(body, &token); err != nil {
+		return nil, fmt.Errorf("Failed to parse json in verify flow.")
+	}
+
+	return token, nil
+}
+
+// Helper struct used in sign JWT
+type signerOpts struct{}
+
+func (r signerOpts) HashFunc() crypto.Hash {
+	return crypto.SHA256
+}
+
+// Base 64 encode a block. The output doesn't contain the trailing double equal
+// signs.
+func base64Encode(b []byte) string {
+	return strings.TrimSuffix(base64.URLEncoding.EncodeToString(b), "==")
+}
+
+// Creates a signed JWT token. Used for two-legged oauth token request.
+func createJWT(secret map[string]interface{}, scopes string, signer signerInterface) (string, error) {
+	// A valid JWT has an "iat" timestamp and an "exp" timestamp. Get the current
+	// time to create these timestamps.
+	now := int(time.Now().Unix())
+
+	// Construct the JWT header, which contains the private key id in the service
+	// account secret.
+	header := map[string]string{
+		"typ": "JWT",
+		"alg": "RS256",
+		"kid": secret["private_key_id"].(string),
+	}
+
+	// Construct the JWT payload.
+	payload := map[string]string{
+		"aud":   secret["token_uri"].(string),
+		"scope": scopes,
+		"iat":   strconv.Itoa(now),
+		"exp":   strconv.Itoa(now + 3600),
+		"iss":   secret["client_email"].(string),
+	}
+
+	// Convert the header and payload map to json.
+	headerJson, err := json.Marshal(header)
+	if err != nil {
+		return "", fmt.Errorf("Failed to convert JWT header to json.")
+	}
+	payloadJson, err := json.Marshal(payload)
+	if err != nil {
+		return "", fmt.Errorf("Failed to convert JWT payload to json.")
+	}
+
+	// Convert the header and payload json to base64
+	headerB64 := base64Encode(headerJson)
+	payloadB64 := base64Encode(payloadJson)
+
+	// The first two segments of the JWT are signed. The signature is the third
+	// segment.
+	segments := headerB64 + "." + payloadB64
+
+	// sign the hash, instead of the actual segments.
+	hash := sha256.New()
+	hash.Write([]byte(segments))
+	hashed := hash.Sum(nil)
+	sOpts := signerOpts{}
+	signedBytes, err := signer.sign(rand.Reader, hashed, crypto.SignerOpts(sOpts))
+	if err != nil {
+		return "", fmt.Errorf("Failed to sign JWT: %s", err)
+	}
+
+	// Generate the final JWT as
+	// base64(header) + "." + base64(payload) + "." + base64(signature)
+	return segments + "." + base64Encode(signedBytes), nil
+}
+
+// Struct for OAuth2 Client
+type Client struct {
+	// The parsed client secret or service account json.
+	secret map[string]interface{}
+
+	// The function to handle two-legged oauth authorization flow. This function
+	// takes in a authorization URL. After the user authorizes on the given url,
+	// returns the verification code.
+	authorizeHandler func(string) string
+}
+
+// Create a new OAuth2 Client
+func NewClient(secretBytes []byte) *Client {
```

Comment:
should be easier for people to use this library, since json file is read as []byte